File: src/nativeMain/kotlin/com/portalsoup/mox/plugin/MoxPlugin.kt
@file:OptIn(ExperimentalForeignApi::class)

package com.portalsoup.mox.plugin

import kotlinx.cinterop.CStructVar
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.NativePtr
import kotlinx.cinterop.objcPtr

//@CStruct("MoxPlugin")
class MoxPlugin(rawPtr: NativePtr): IMoxPlugin, CStructVar(rawPtr) {
    override val name: String
        get() = TODO("Not yet implemented")
    override val parentName: String?
        get() = TODO("Not yet implemented")

    override fun run(args: String) = this.objcPtr()

//    companion object : Type(sizeOf<MoxPlugin>(), alignOf<MoxPlugin>())
}

File: example-c-plugin/mox_plugin.c
#include "../src/nativeInterop/cinterop/mox_plugin.h"
#include <stdio.h>

static void run(const char* args) {
    printf("[Plugin called with] %s", args);
}

static Plugin plugin = {
    .name = "c_hello",
    .run = run
};

Plugin* createPlugin(void) {
    return &plugin;
}